File: firestorm/src/main/java/com/kazzla/core/async/io/IO.java
/*
 * Copyright (c) 2013 koiroha.org.
 * All sources and related resources are available under Apache License 2.0.
 * http://www.apache.org/licenses/LICENSE-2.0.html
*/
package com.kazzla.core.io;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.nio.charset.Charset;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.util.UUID;

// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// IO
// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
public final class IO {
	private static final Logger logger = LoggerFactory.getLogger(IO.class);

	public static final Charset UTF8 = Charset.forName("UTF-8");

	// ==============================================================================================
	// コンストラクタ
	// ==============================================================================================
	/**
	 * コンストラクタはクラス内に隠蔽されています。
	 */
	private IO() { }

	// ==============================================================================================
	// リソースのクローズ
	// ==============================================================================================
	/**
	 * 指定されたリソースを例外なしでクローズします。
	 * @param cs クローズするリソース
	 */
	public static void close(Closeable... cs){
		for(Closeable c: cs){
			try {
				if(c != null){
					c.close();
				}
			} catch(IOException ex){
				logger.warn("fail to destroy object", ex);
			}
		}
	}

	// ==============================================================================================
	// メッセージダイジェストの参照
	// ==============================================================================================
	/**
	 * 指定されたアルゴリズムを使用して入力ストリームから取得できるデータのメッセージダイジェストを参照します。
	 */
	public static byte[] getMessageDigest(InputStream is, String algorithm) throws IOException {
		try {
			MessageDigest md = MessageDigest.getInstance(algorithm);
			DigestInputStream in = new DigestInputStream(is, md);
			byte[] buffer = new byte[1024];
			while(in.read(buffer) > 0);
			return md.digest();
		} catch(Exception ex){
			throw new IOException(ex);
		}
	}

}

File: share/src/main/scala/com/kazzla/io.scala
/*
 * Copyright (c) 2013 koiroha.org.
 * All sources and related resources are available under Apache License 2.0.
 * http://www.apache.org/licenses/LICENSE-2.0.html
*/
package com.kazzla

import java.io._
import java.net.{InetSocketAddress, SocketAddress}
import java.security.{DigestInputStream, MessageDigest}
import java.util.UUID
import org.slf4j._
import scala.annotation.tailrec

// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
// io
// ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
/**
 * @author Takami Torao
 */
package object io {

	import scala.language.reflectiveCalls

	private[this] val logger = LoggerFactory.getLogger("com.kazzla.io")

	// ==============================================================================================
	// クローズ
	// ==============================================================================================
	/**
	 * `close()` メソッドを実装しているオブジェクトに対して例外なしでクローズ操作をおこないます。
	 */
	def close[T <% { def close():Unit }](cs:T*):Unit = cs.filter{ _ != null }.foreach { c =>
		try {
			c.close()
		} catch {
			case ex:IOException =>
				logger.warn(s"fail to close resource: $c", ex)
		}
	}

	// ==============================================================================================
	// リソースのスコープ設定
	// ==============================================================================================
	/**
	 * 指定されたクローズ可能なオブジェクトに対してラムダ実行後にクローズの実行を保証します。
	 */
	def using[T <% { def close():Unit }, U](resource:T)(f:(T)=>U):U = try {
		f(resource)
	} finally {
		close(resource)
	}

	// ==============================================================================================
	// リソースのスコープ設定
	// ==============================================================================================
	/**
	 * 指定されたクローズ可能なオブジェクトに対してラムダ実行後にクローズの実行を保証します。
	 */
	def usingOutput[T](file:File)(f:(OutputStream)=>T):T = using(new FileOutputStream(file))(f)
	def usingInput[T](file:File)(f:(InputStream)=>T):T = using(new FileInputStream(file))(f)

	// ==============================================================================================
	// ストリームのコピー
	// ==============================================================================================
	/**
	 * 指定されたバッファを用いてストリームのコピーを行います。
	 */
	@tailrec
	def copy[T <% { def write(b:Array[Byte],o:Int,l:Int):Unit }](src:InputStream, dst:T, buffer:Array[Byte], length:Int = Int.MaxValue):Unit = {
		if(length > 0){
			val len = src.read(buffer, 0, math.min(buffer.length, length))
			if(len > 0){
				dst.write(buffer, 0, len)
				copy(src, dst, buffer, length - len)
			}
		}
	}

	def copy[T <% { def write(b:Array[Byte],o:Int,l:Int):Unit }](src:InputStream, dst:T):Unit = {
		copy(src, dst, new Array[Byte](1024))
	}

	implicit class RichSocketAddress(addr:SocketAddress) {
		def getName:String = addr match {
			case i:InetSocketAddress =>
				s"${i.getAddress.getHostAddress}:${i.getPort}"
			case s => s.toString
		}
	}

	implicit class RichInputStream(in:InputStream) {

		/**
		 * メッセージダイジェストの算出
		 */
		def digest(algorithm:String):Array[Byte] = {
			val md = MessageDigest.getInstance(algorithm)
			val is = new DigestInputStream(in, md)
			val buffer = new Array[Byte](1024)
			while(is.read(buffer) > 0){
				None
			}
			md.digest()
		}

		/**
		 *
		 */
		def readFully():Array[Byte] = {
			val out = new ByteArrayOutputStream()
			val buffer = new Array[Byte](1024)
			copy(in, out, buffer)
			out.toByteArray
		}

		/**
		 *
		 */
		def readText(enc:String):String = new String(readFully(), enc)

	}

	/**
	 *
	 */
	implicit class RichDataOutputStream(out:DataOutputStream){
		def writeUUID(uuid:UUID):Unit = {
			out.writeLong(uuid.getMostSignificantBits)
			out.writeLong(uuid.getLeastSignificantBits)
		}
	}

	implicit class RByteArray(array:Array[Byte]){
		def toHexString:String = {
			array.map{ b => f"${b & 0xFF}%02X" }.mkString
		}
	}

}

